src/ptvsd/adapter/debuggee.py

```diff
@@ -461,3 +457,45 @@ def _pid_to_handle(pid):
         if not handle:
             raise ctypes.WinError()
         return handle
+
+
+class ProcessPidServer(object):
+ 
+    def __init__(self):
+        self.listener = socket.create_server("127.0.0.1", 0)
+        self.host, self.port = self.listener.getsockname()
+        self.wait_thread = threading.Thread(
+            target=self._wait_for_pid, name="Process Pid Server"
+        )
+        self.wait_thread.daemon = True
+        self.wait_thread.start()
+
+    def _wait_for_pid(self):
+        try:
+            log.info(
+                "Adapter waiting for connection from launcher on {0}:{1}...",
+                self.host,
+                self.port
+            )
+            sock, (l_host, l_port) = self.listener.accept()
+        finally:
+            self.listener.close()
+        log.info("Launcher connection accepted from {0}:{1}.", l_host, l_port)
+
+        data = b''
+        with sock:
```

Comment:
I think you can skip the loop if you do `makefile(sock).read()` - this should give you everything that was sent right up until the socket gets closed. This way, you can also specify text mode for `makefile()`, and don't need to decode anything.